### ISSUE TITLE
freenas-api-smb does not use fsGroup config

### ISSUE THREAD

**Issue by eaglesemanation** [democratic-csi/democratic-csi#432]
Hi, I've been using `freenas-api-iscsi` for a while now, great piece of software, quite reliable (outside of read-only after reboot shenanigans)

Now I wanted to have a share that I could access outside k8s pod, because it needs to be frequently modified, and I want to skip whole GitOps paradigm for this, at least for now. So I tried using `freenas-api-smb`. The issue is that whatever I do, I can't make it mount as anything other than root. One thing of note - I'm nesting an SMB mount under iSCSI mount within pod, but I don't think that would affect anything, would it?

Here is my full config and a PVC example:

<details open>
<summary>HelmRelease</summary>

```yaml
# yaml-language-server: $schema=https://raw.githubusercontent.com/datreeio/CRDs-catalog/main/helm.toolkit.fluxcd.io/helmrelease_v2.json
apiVersion: helm.toolkit.fluxcd.io/v2
kind: HelmRelease
metadata:
  name: democratic-csi-smb
  namespace: democratic-csi
spec:
  interval: 30m
  releaseName: democratic-csi-smb
  chart:
    spec:
      chart: democratic-csi
      version: 0.14.7
      sourceRef:
        kind: HelmRepository
        name: democratic-csi
        namespace: flux-system
      interval: 30m
  dependsOn:
    - name: snapshot-controller
      namespace: external-snapshotter
  values:
    controller:
      driver:
        image: docker.io/democraticcsi/democratic-csi:v1.9.3
        resources:
          requests:
            cpu: 20m
            memory: 128Mi
          limits:
            memory: 512Mi
    node:
      driver:
        image: docker.io/democraticcsi/democratic-csi:v1.9.3
        resources:
          requests:
            cpu: 20m
            memory: 128Mi
          limits:
            memory: 256Mi
    csiDriver:
      name: "org.democratic-csi.freenas-api-smb"
      fsGroupPolicy: File # Just added it, doesn't seem to fix this
    volumeSnapshotClasses:
      - name: freenas-smb-csi
        labels:
          velero.io/csi-volumesnapshot-class: "true"
        parameters:
          detachedSnapshots: "true"
    driver:
      existingConfigSecret: freenas-api-smb-driver
      config:
        driver: freenas-api-smb

```

</details>

<details open>
<summary>Driver config secret</summary>

```yaml
apiVersion: v1
kind: Secret
metadata:
    name: freenas-api-smb-driver
    namespace: democratic-csi
stringData:
    driver-config-file.yaml: |-
        driver: freenas-api-smb
        instance_id: null
        httpConnection:
          protocol: https
          host: ${NAS_ADDR} # FluxCD automatically replaces this
          port: 443
          apiKey: <redacted>
          allowInsecure: true
        zfs:
          datasetProperties:
            "org.freenas:description": "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}/{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
          
          datasetParentName: erie/csi/smb/vols
          detachedSnapshotsDatasetParentName: erie/csi/smb/snaps
          datasetEnableQuotas: true
          datasetEnableReservation: false
          datasetPermissionsMode: "0777"
          datasetPermissionsUser: 3001 # UID of user on TrueNAS that is being used to log into SMB
          datasetPermissionsGroup: 0
        smb:
          shareHost: ${NAS_ADDR}
          nameTemplate: ""
          namePrefix: ""
          nameSuffix: ""
          shareHome: false
          shareAllowedHosts: []
          shareDeniedHosts: []
          #shareDefaultPermissions: true
          shareGuestOk: true
          #shareGuestOnly: true
          #shareShowHiddenFiles: true
          shareRecycleBin: true
          shareBrowsable: false
          shareAccessBasedEnumeration: true
          shareTimeMachine: false
          #shareStorageTask: null
          dirPermissionsMode: "0777" # Just added it while debugging this issue

```

</details>

<details open>
<summary>Storage Class (separate from helm so I could add node-stage-secret)</summary>

```yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: freenas-api-smb-csi
  labels:
    app.kubernetes.io/name: democratic-csi
    app.kubernetes.io/instance: democratic-csi-smb
provisioner: org.democratic-csi.freenas-api-smb
reclaimPolicy: Delete
allowVolumeExpansion: true
volumeBindingMode: Immediate
parameters:
  csi.storage.k8s.io/fstype: cifs
  csi.storage.k8s.io/node-stage-secret-name: freenas-smb-creds
  csi.storage.k8s.io/node-stage-secret-namespace: democratic-csi
```

</details>

<details open>
<summary>Node Stage Secret</summary>

```yaml
apiVersion: v1
kind: Secret
metadata:
    name: freenas-smb-creds
    namespace: democratic-csi
stringData:
    mount_flags: username=<redacted>,password=<redacted>

```

</details>

<details open>
<summary>Example PVC</summary>

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mc-config
  namespace: minecraft
spec:
  accessModes:
    - ReadWriteOnce
  volumeMode: Filesystem
  resources:
    requests:
      storage: 1Gi
  storageClassName: freenas-api-smb-csi
```

</details>

<details open>
<summary>Example Deployment</summary>

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: minecraft-mc-minecraft
  namespace: minecraft
spec:
  replicas: 1
  selector:
    matchLabels:
      app: minecraft-mc-minecraft
  strategy:
    type: Recreate
  template:
    metadata:
      labels:
        app: minecraft-mc-minecraft
    spec:
      containers:
      - env:
        - <skipped>
        image: itzg/minecraft-server:latest
        imagePullPolicy: IfNotPresent
        name: minecraft-mc-minecraft
        ports:
        - containerPort: 25565
          name: minecraft
          protocol: TCP
        resources:
          requests:
            cpu: "1"
            memory: 4Gi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            drop:
            - ALL
          readOnlyRootFilesystem: true
        volumeMounts:
        - mountPath: /tmp
          name: tmp
        - mountPath: /data
          name: datadir
        - mountPath: /backups
          name: backupdir
          readOnly: true
        - mountPath: /data/config
          name: config
      securityContext:
        fsGroup: 2000
        fsGroupChangePolicy: Always
        runAsGroup: 3000
        runAsNonRoot: true
        runAsUser: 1000
        seccompProfile:
          type: RuntimeDefault
      volumes:
      - emptyDir: {}
        name: tmp
      - name: datadir
        persistentVolumeClaim:
          claimName: minecraft-mc-minecraft-datadir # freenas-api-iscsi volume
      - emptyDir: {}
        name: backupdir
      - name: config
        persistentVolumeClaim:
          claimName: maria-mc-config
```

</details>

<details open>
<summary>Viewing permissions within pod</summary>

```
minecraft@minecraft-mc-minecraft-85d9fc68bb-qdq5f:~$ ls -lhd /data/config
drwxr-xr-x 2 root root 0 Nov 30 09:56 /data/config
minecraft@minecraft-mc-minecraft-85d9fc68bb-qdq5f:~$ ls -lhd /data/
drwxrwsr-x 5 root 2000 4.0K Nov 30 20:34 /data/
```

</details>

<details open>
<summary>Viewing permissions outside pod</summary>

```
> talosctl -n <pod node ip> ls -l /var/lib/kubelet/plugins/kubernetes.io/csi/org.democratic-csi.freenas-api-smb/5ba2a24b53fb79466e45b8b79e1223a607463eddd006c6e478348a93b54253ba/globalmount
NODE             MODE         UID   GID   SIZE(B)   LASTMOD           LABEL   NAME
192.168.25.114   drwxr-xr-x   0     0     0         Nov 30 04:56:03           .

> talosctl -n <pod node ip> ls -l /var/lib/kubelet/pods/f24da2cd-1364-4b39-94d0-d07a2e3a9a3b/volumes/kubernetes.io~csi/pvc-1512d79f-5abb-4c80-ae6f-b1109ef7769a/mount
NODE             MODE         UID   GID   SIZE(B)   LASTMOD           LABEL   NAME
192.168.25.114   drwxr-xr-x   0     0     0         Nov 30 04:56:03           .
```

</details>

For now I'll probably just run the pod as root, but I'd like to fix this properly

**Comment by travisghansen** [comment]
Welcome! You almost certainly need `fsGroupPolicy: File # Just added it, doesn't seem to fix this`. I think the nested bits may be causing the issue but not entirely sure. This feature isn't actually implemented by the csi drivers themselves but rather kubelet has the job of invoking `chown` on the mounted directory on the host.

**Comment by eaglesemanation** [comment]
Finally had a reason to try using SMB driver again, and got the same issue. Now I had a bit more experience debugging k8s, so I found out that kubelet was indeed running chown+chmod on mounted volumes "successfully". But when I ran `ls -la` from within the container - group was still `0`. Then I tried running the container as root, and running `chown root:1000 <dirname>`, and it would exit with status code 0, but the ownership remained `0:0`.
So after searching for similar issues with CIFS mount I found this: https://superuser.com/questions/744384/allow-chmod-on-cifs-mount

TLDR: had to add `noperm` to mount_flags.
Feel free to close this, not closing myself in case this is worthy of documentation / examples change
